plugins/InspectionGadgets/InspectionGadgetsAnalysis/src/com/siyeh/ig/bugs/BaseEqualsVisitor.java

```diff
@@ -53,19 +67,156 @@ public void visitMethodReferenceExpression(PsiMethodReferenceExpression expressi
+        return getType(argument);
+      }
+    }
+    else if (parent instanceof PsiExpressionList) {
+      return findPredicateTypeFromExpressionList(upperPredicate, (PsiExpressionList)parent);
+    }
+    return null;
+  }
+
+  @Nullable
+  private static PsiType findPredicateTypeFromExpressionList(PsiExpression upperPredicate, PsiExpressionList expressionList) {
+    Integer index = findParameterIndex(expressionList, upperPredicate);
+    if (index == null || expressionList.getExpressionCount() <= index) {
+      return null;
+    }
+
+    if (!(expressionList.getParent() instanceof PsiMethodCallExpression)) {
+      return null;
+    }
+
+    final PsiMethodCallExpression expectedCall = (PsiMethodCallExpression)expressionList.getParent();
+    PsiMethod method = expectedCall.resolveMethod();
+    if (method == null) {
+      return null;
+    }
+    PsiType[] arguments = method.getSignature(expectedCall.resolveMethodGenerics().getSubstitutor())
+      .getParameterTypes();
+
+    if (arguments.length <= index) {
+      return null;
+    }
+    PsiType type = arguments[index];
+    if (!(type instanceof PsiClassType)) {
+      return null;
+    }
+    PsiClassType classType = (PsiClassType)type;
+    if (classType.getParameterCount() != 1 || !PsiTypesUtil.classNameEquals(classType, CommonClassNames.JAVA_UTIL_FUNCTION_PREDICATE)) {
+      return null;
+    }
+    PsiType parameter = classType.getParameters()[0];
+    //Object can be cast to any
+    if (parameter == null || parameter.equalsToText(CommonClassNames.JAVA_LANG_OBJECT)) {
+      return null;
+    }
+    if (parameter instanceof PsiWildcardType) {
+      PsiWildcardType psiWildcardType = (PsiWildcardType)parameter;
+      parameter = psiWildcardType.isSuper() ? psiWildcardType.getSuperBound() : parameter;
+    }
+    return parameter;
+  }
+
+  @Nullable
+  private static Integer findParameterIndex(PsiExpressionList list, PsiExpression predicate) {
+    for (int i = 0; i < list.getExpressionCount(); i++) {
+      final PsiExpression expression = list.getExpressions()[i];
+      if (predicate.equals(PsiUtil.skipParenthesizedExprDown(expression))) {
+        return i;
+      }
+    }
+    return null;
+  }
+
+  @Nullable
+  private static PsiType findPredicateParameter(PsiExpression predicate) {
+    final PsiType returnType = predicate.getType();
+    if (!(returnType instanceof PsiClassType)) {
+      return null;
+    }
+
+    PsiClassType classType = (PsiClassType)returnType;
+    if (classType.getParameterCount() != 1) {
+      return null;
+    }
+    PsiType parameter = classType.getParameters()[0];
+    if (!parameter.equalsToText(CommonClassNames.JAVA_LANG_OBJECT)) {
+      return parameter;
+    }
+    return null;
+  }
+
+  private static boolean isPredicateSourceForIsEqual(PsiElement parent) {
+    return parent instanceof PsiMethodCallExpression &&
+           PREDICATE_SOURCE_FOR_FIND_IS_EQUAL.test((PsiMethodCallExpression)parent);
```

Comment:
You can use `.matches` instead of `.test` to avoid instanceof+cast. In this case, the method could be inlined.

Reply:
Oh. Cool! Thanks! Inline method and rewrite logic of "resolveIsEqualPredicateType" method